src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,315 @@
+// LinkedList uses nulls, so we need to disable the nullability warnings for this file.
+namespace FSharp.Compiler.Caches
+
+open System
+open System.Collections.Generic
+open System.Collections.Concurrent
+open System.Threading
+open System.Diagnostics
+open System.Diagnostics.Metrics
+
+[<Struct; RequireQualifiedAccess; NoComparison; NoEquality>]
+type CacheOptions =
+    {
+        /// Total capacity, determines the size of the underlying store.
+        TotalCapacity: int
+
+        /// Safety margin size as a percentage of TotalCapacity.
+        HeadroomPercentage: int
+    }
+
+    static member Default =
+        {
+            TotalCapacity = 128
+            HeadroomPercentage = 50
+        }
+
+[<Sealed; NoComparison; NoEquality>]
+[<DebuggerDisplay("{ToString()}")>]
+type CachedEntity<'Key, 'Value> =
+    val mutable Key: 'Key
+    val mutable Value: 'Value
+    val mutable AccessCount: int64
+    val mutable Node: LinkedListNode<CachedEntity<'Key, 'Value>>
+
+    new(key, value) =
+        {
+            Key = key
+            Value = value
+            AccessCount = 0L
+            Node = Unchecked.defaultof<_>
+        }
+
+    member this.WithNode() =
+        this.Node <- LinkedListNode(this)
+        this
+
+    member this.ReUse(key, value) =
+        this.Key <- key
+        this.Value <- value
+        this.AccessCount <- 0L
+        this
```

Comment:
What should be doen with `Node` at reuse time?

Reply:
Don't touch it. I will add a comment to make clear the idea. It is important that the Node holds the entity, and the entity knows it's node. Once we assigned it, it's good forever, even when reused.